enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,893 @@
+kind: ContainerRuntimeConfig
+metadata:
+  ...
+spec:
+  containerRuntimeConfig:
+    additionalImageStores:
+    - /var/lib/upgrade/4.13.7-x86_64
+```
+
+When the new `additionalImageStores` field is added or changed the machine
+config operator will need to regenerate the `/etc/containers/storage.conf` file,
+including the corresponding `additionalimagestores` field, and then reload it
+using the equivalent of `systemctl reload crio.service`.
+
+The changes to pin the images will be done in a
+`/etc/crio/crio.conf.d/pin-upgrade.conf` file, something like this:
+
+```toml
+pinned_images=[
+  "quay.io/openshift-release-dev/ocp-release@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  ...
+]
+```
+
+These files aren't tracked by the machine config operator, but the CRI-O service
+needs to be reloaded when they are created or changed. To support that a new
+filed will be added to the `ContainerRuntimeConfig` object, for example:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+  name: pin-upgrade
+spec:
+  containerRuntimeConfig:
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    ...
+```
+
+When the new `pinnedImages` field is added or changed the machine config
+operator will need to create or update the corresponding
+`/etc/crio/crio.conf.d/pin-upgrade.conf` file, and then reload it using the
+equivalent of `systemctl reload crio.service`.
+
+#### Tool to create the upgrade bundle
+
+The upgrade bundle will be created by an engineer in a partner's factory using a
+new `oc adm upgrade create bundle` command. This engineer will first determine
+the target version number, for example 4.13.7. Note that doing this will
+probably require access to the upgrade service available in api.openshift.com.
+Finding that upgrade version number is outside of the scope of this enhancement.
+
+The engineer will then need internet access and a Linux machine where she can
+run `oc adm upgrade create bundle`, for example:
+
+```bash
+$ oc adm upgrade create bundle \
+--arch=x86_64 \
+--version=4.13.7 \
+--pull-secret=/my/pull/secret.txt \
+--output=/my/bundle/dir
+```
+
+The `oc adm upgrade bundle` command will find the list of image references that
+make up the release, doing the equivalent of this:
+
+```bash
+$ oc adm release info \
+quay.io/openshift-release-dev/ocp-release:4.13.7-x86_64 -o json | \
+jq '.references.spec.tags[].from.name'
+```
+
+In addition to the release images the tool will also support explicitly adding
+custom images. For example:
+
+```bash
+$ oc adm upgrade create bundle \
+--arch=x86_64 \
+--version=4.13.7 \
+--pull-secret=/my/pull/secret.txt \
+--extra-image=quay.io/my-company/my-workload1 \
```

Comment:
Passing a file with a list of image references would make it easier to avoid command line length limits that might result if there are a lot of images.

Reply:
Makes sense, added a `--extra-images-file` option.